nautobot/docs/development/core/model-checklist.md

```diff
@@ -0,0 +1,139 @@
+- ▢ Add field to `<Model>FilterForm.fields`
+- ▢ Add field column to `<Model>Table.fields`
+    - ▢ Keep `"pk"` as the first column and `"actions"` (if present) as the last column
+    - ▢ _optional_ Add field column to `<Model>Table.default_columns` if desired
+- ▢ Add field to `nautobot/<app>/templates/<app>/<model>.html` detail template
+- ▢ Add field to test data in appropriate unit tests (including view, filter, model, and API tests)
+    - ▢ Add field testing in`test_api.<Model>TestCase`
+    - ▢ Add field testing in `test_filters.<Model>TestCase`
+    - ▢ Add field testing in `test_models.<Model>TestCase`
+    - ▢ Add field testing in `test_views.<Model>TestCase`
+    - ▢ _optional_ Add field testing in `test_forms.<Model>TestCase` if applicable
+- ▢ Validate that the field appears correctly in GraphQL
+    - ▢ If the field is not compatible with GraphQL or shouldn't be included in GraphQL it's possible to exclude a specific field in the GraphQL Type Object associated with this specific model. You can refer to the [graphene-django documentation](https://docs.graphene-python.org/projects/django/en/latest/queries/#specifying-which-fields-to-include) for additional information.
+- ▢ Add field to model documentation with an appropriate version annotation
+- ▢ _optional_ Add field to `<Model>Serializer` if default representation isn't optimal
+- ▢ _optional_ Add field to `<Model>Serializer.fields` if it's not using `fields = ["__all__"]`
+- ▢ _optional_ Add field to any custom create/edit template
+- ▢ _optional_ Update model `clean()` with any new logic needed
+
+### Adding a Foreign Key from ModelA to ModelB
+
+- ▢ Select appropriate `on_delete` behavior (`SET_NULL`, `CASCADE`, `PROTECT`)
+- ▢ Add a `DynamicModelChoiceField` to `<ModelA>Form`
+- ▢ Add a `NaturalKeyOrPKMultipleChoiceFilter` to `<ModelA>FilterSet`
+- ▢ Add a `DynamicModelMultipleChoiceField(required=False, ...)` to `<ModelA>FilterForm`
+- ▢ Add a `DynamicModelChoiceField(required=False, ...)` to `<ModelA>BulkEditForm`
+- ▢ Add `select_related` to API `<ModelA>ViewSet.queryset`
+- ▢ Add `select_related` to UI `<ModelA>UIViewSet.queryset`
+- ▢ _optional_ Add `annotate(<model_a>_count=count_related(ModelA,...)` to `<ModelB>ViewSet.queryset`
+    - ▢ Add `<model_a>_count` field to `<ModelB>Serializer`
+- ▢ _optional_ Add `<model_a>_count = LinkedCountColumn(...)` to `<ModelB>Table` and `<ModelB>Table.fields`
+- ▢ _optional_ Add related object table to ModelB detail view
+
+#### Adding a Status or Role field to a Model
+
+- ▢ Use `StatusField` or `RoleField` in place of a generic `ForeignKey`
+- ▢ Add [data migration](https://docs.djangoproject.com/en/stable/topics/migrations/) providing default Status/Role records for this model
+    - ▢ Remember: data migrations must not share a file with schema migrations or vice versa!
+    - ▢ Specify appropriate migration dependencies (for example, against another app that you're using models from)
```

Comment:
We also should probably mention that we need to update the `role_retrieve.html` while adding a role field to a model. Since we have a test in place to check whether the related table is rendered in that template. `nautobot.extras.tests.test_views.RoleTestCase.test_view_with_content_types`

Reply:
Does this fall within the general guidance at line 113 above (`_optional_ Add related object table to ModelB detail view`), or is this something we need to call out separately?

Reply:
I think it is really not `optional` since we have a test in place to check the presence of related object tables. I would say it is mandatory for role-related Models at least so it might be worth calling out separately IMHO.

Reply:
Fair, will add.